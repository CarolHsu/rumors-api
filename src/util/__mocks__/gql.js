/**
 * Usage:
 *
 * jest.mock('path-to-gql/gql');
 *
 * it('some test', async () => {
 *   gql.__addMockResponse({data: {queryName: result})
 *
 *   const result = await functionUnderTestThatInvokesGql();
 *
 *   expect(gql.__getRequests).toMatchSnapshot(); // assert request sent
 *   expect(result).toMatchSnapshot();
 *
 *   gql.__reset() // Must reset to reset mock responses & call sequence number!
 * })
 */

let seq = 0;
let mockResponses = [];
let requests = [];

function gql(query, ...substitutions) {
  if (!mockResponses[seq]) {
    throw Error(
      `gql Mock error: No response found for request #${seq}. Please add mock response first.`
    );
  }

  return variables => {
    const queryAndVariable = {
      query: String.raw(query, ...substitutions),
    };

    if (variables) queryAndVariable.variables = variables;

    requests.push(queryAndVariable);

    return Promise.resolve(mockResponses[seq++]);
  };
}

gql.__addMockResponse = resp => mockResponses.push(resp);
gql.__getRequests = () => requests;
gql.__reset = () => {
  seq = 0;
  mockResponses = [];
  requests = [];
};

export default gql;
