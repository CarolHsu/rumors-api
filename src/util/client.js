import config from 'config';
import elasticsearch from 'elasticsearch';

export default new elasticsearch.Client({
  host: config.get('ELASTICSEARCH_URL'),
  log: 'trace',
});
