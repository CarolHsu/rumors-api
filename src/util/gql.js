import fetch from 'node-fetch';
import rollbar from '../rollbarInstance';
import url from 'url';
import config from 'config';

const API_URL = config.get('URL_RESOLVER_URL') || 'http://localhost:4000/';

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
// GraphQL Protocol: http://dev.apollodata.com/tools/graphql-server/requests.html
//
export default (query, ...substitutions) => (variables, search) => {
  const queryAndVariable = {
    query: String.raw(query, ...substitutions),
  };

  if (variables) queryAndVariable.variables = variables;

  let status;
  const URL = `${API_URL}${url.format({ query: search })}`;

  return fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(queryAndVariable),
  })
    .then(r => {
      status = r.status;
      return r.json();
    })
    .then(resp => {
      if (status === 400) {
        throw new Error(
          `GraphQL Error: ${resp.errors
            .map(({ message }) => message)
            .join('\n')}`
        );
      }
      if (resp.errors) {
        // When status is 200 but have error, just print them out.
        // eslint-disable-next-line no-console
        console.error('GraphQL operation contains error:', resp.errors);
        rollbar.error(
          'GraphQL error',
          {
            body: JSON.stringify(queryAndVariable),
            url: URL,
          },
          { resp }
        );
      }
      return resp;
    });
};
