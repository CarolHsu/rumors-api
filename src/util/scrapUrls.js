import rollbar from 'rollbarInstance';
import urlRegex from 'url-regex';
import DataLoader from 'dataloader';

import gql from './gql';

/**
 * Extracts urls from a string.
 * Fetches the specified url data from cache or through scrapping.
 * Optionally writes the scrapped result to database.
 *
 * @param {string} text
 * @param {object} options
 * @param {object} options.cacheLoader - The dataloader that loads result from `urls`, given a URL.
 *                                       If not given, scrapUrls don't read `urls` cache.
 * @param {boolean} options.noFetch - If true, return null when cache does not hit. Default: false.
 * @param {object} options.client - ElasticSearch.js client instance used to write new results to
 *                                  `urls` index. If not given, scrapUrl don't write `urls` cache.
 * @return {Promise<ScrapResult[]>} - If cannot scrap, resolves to null
 */
async function scrapUrls(text, { cacheLoader, client, noFetch = false } = {}) {
  const urls = text.match(urlRegex()) || [];
  if (urls.length === 0) return [];

  const scrapLoader = new DataLoader(urls =>
    gql`
      query($urls: [String]!) {
        resolvedUrls(urls: $urls) {
          url
          canonical
          title
          summary
          topImageUrl
          html
          status
        }
      }
    `({ urls }).then(({ data }) => data.resolvedUrls)
  );

  const result = await Promise.all(
    // 1st pass: resolve cache
    urls.map(async url => {
      if (cacheLoader) {
        const result = await cacheLoader.load(url);

        if (result)
          return {
            ...result,
            url, // Overwrite the URL from cache with the actual url in text, because cacheLoader may match canonical URLs
            fromUrlsCache: true,
          };
      }

      return url;
    })
  ).then(results => {
    // 2nd pass: scrap when needed
    return Promise.all(
      results.map(async result => {
        if (typeof result !== 'string') {
          return result;
        }

        // result is an URL here

        if (noFetch) return null;

        return scrapLoader.load(result);
      })
    );
  });

  if (!client) {
    // client not specified, don't to write to urls
    return result;
  }

  const fetchedAt = new Date();

  // Filter out null, write to "urls" index
  const urlsBody = result.reduce((body, r) => {
    if (!r || r.fromUrlsCache) {
      return body;
    }

    const { url, canonical, title, summary, topImageUrl, html, status } = r;

    return body.concat([
      { index: { _index: 'urls', _type: 'doc' } },
      {
        canonical,
        title,
        summary,
        topImageUrl,
        html,
        url,
        fetchedAt,
        status,
      },
    ]);
  }, []);

  if (urlsBody.length === 0) {
    // Nothing to write to urls
    return result;
  }

  const insertResult = await client.bulk({
    body: urlsBody,
  });

  if (insertResult.errors) {
    rollbar.error('Urls insert error', insertResult);
  }

  return result;
}

export default scrapUrls;
