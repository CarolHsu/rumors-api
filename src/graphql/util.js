import {
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLFloat,
} from 'graphql';

import client from 'util/client';

// Deprecated.
//
export function scoredDocFactory(name, type) {
  return new GraphQLObjectType({
    name,
    fields: {
      score: { type: GraphQLFloat },
      doc: { type },
    },
  });
}

// https://www.graph.cool/docs/tutorials/designing-powerful-apis-with-graphql-query-parameters-aing7uech3
//
// Filtering args definition & parsing
//

export function getArithmeticExpressionType(typeName, argType) {
  return new GraphQLInputObjectType({
    name: typeName,
    fields: {
      LT: { type: argType },
      GT: { type: argType },
      EQ: { type: argType },
    },
  });
}

export function getOperatorAndOperand(expression) {
  if (typeof expression.EQ !== 'undefined') {
    return { operator: '=', operand: expression.EQ };
  } else if (typeof expression.LT !== 'undefined') {
    return { operator: '<', operand: expression.LT };
  } else if (typeof expression.GT !== 'undefined') {
    return { operator: '>', operand: expression.GT };
  }
  return {};
}

export function createFilterType(typeName, args) {
  const filterType = new GraphQLInputObjectType({
    name: typeName,
    fields: () => ({
      ...args,
      // TODO: converting nested AND / OR to elasticsearch
      // AND: { type: new GraphQLList(filterType) },
      // OR: { type: new GraphQLList(filterType) },
    }),
  });
  return filterType;
}

//
// Sort args definition & parsing
//

const SortOrderEnum = new GraphQLEnumType({
  name: 'SortOrderEnum',
  values: {
    ASC: { value: 'asc' },
    DESC: { value: 'desc' },
  },
});

export function createSortType(typeName, filterableFieldNames = []) {
  return new GraphQLList(new GraphQLInputObjectType({
    name: typeName,
    description: 'An entry of orderBy argument. Specifies field name and the sort order. Only one field name is allowd per entry.',
    fields: filterableFieldNames.reduce(
      (fields, fieldName) => ({ ...fields, [fieldName]: { type: SortOrderEnum } }),
      {},
    ),
  }));
}

export const pagingArgs = {
  first: {
    type: GraphQLInt,
    description: 'Returns only first <first> results',
    defaultValue: 10,
  },
  after: {
    type: GraphQLString,
    description: 'Returns result after this cursor',
  },
};

export function getSortArgs(orderBy, fieldFnMap = {}) {
  return orderBy
    .map((item) => {
      const field = Object.keys(item)[0];
      const order = item[field];
      const defaultFieldFn = o => ({ [field]: { order: o } });

      return (fieldFnMap[field] || defaultFieldFn)(order);
    }).concat({ _uid: { order: 'desc' } }); // enforce at least 1 sort order for pagination
}

// Export for custom resolveEdges() and resolvePageInfo()
//
export function getCursor(cursor) {
  return Buffer.from(JSON.stringify(cursor)).toString('base64');
}

export function getSearchAfterFromCursor(cursor) {
  return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
}

// All search
//
export function createConnectionType(
  typeName,
  nodeType,
  {
    // Default resolvers
    //

    resolveTotalCount = async searchContext =>
      // .count() does not accept sort & size
       (await client.count({
         ...searchContext,
         body: {
           ...searchContext.body,
           size: undefined,
           sort: undefined,
         },
       })).count,

    resolveEdges = async (searchContext, args, { loaders }) => {
      const nodes = await loaders.searchResultLoader.load(searchContext);
      return nodes.map(({ _score: score, _cursor, ...node }) => ({
        node, cursor: getCursor(_cursor), score,
      }));
    },

    resolvePageInfo = async (searchContext, args, { loaders }) => {
      // sort: [{fieldName: {order: 'desc'}}, {fieldName2: {order: 'desc'}}, ...]
      //
      const newSort = searchContext.body.sort.map((item) => {
        const field = Object.keys(item)[0];
        const order = item[field].order === 'desc' ? 'asc' : 'desc';
        return {
          [field]: {
            ...item[field],
            order,
          },
        };
      });

      const lastNode = (await loaders.searchResultLoader.load({
        ...searchContext,
        body: {
          ...searchContext.body,
          sort: newSort,
        },
        size: 1,
      }))[0];

      return {
        lastCursor: getCursor(lastNode._cursor),
      };
    },
  } = {},
) {
  return new GraphQLObjectType({
    name: typeName,
    fields: {
      totalCount: {
        type: GraphQLInt,
        resolve: resolveTotalCount,
      },
      edges: {
        type: new GraphQLList(new GraphQLObjectType({
          name: `${typeName}Edges`,
          fields: {
            node: { type: nodeType },
            cursor: { type: GraphQLString },
            score: { type: GraphQLFloat },
          },
        })),
        resolve: resolveEdges,
      },
      pageInfo: {
        type: new GraphQLObjectType({
          name: `${typeName}PageInfo`,
          fields: {
            lastCursor: { type: GraphQLString },
          },
        }),
        resolve: resolvePageInfo,
      },
    },
  });
}
