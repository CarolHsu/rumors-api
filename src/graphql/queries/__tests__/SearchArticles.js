import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import fixtures from '../__fixtures__/SearchArticle';

describe('SearchArticles', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('fetches related articles', async () => {
    expect(
      await gql`{
      SearchArticles(
        text: "憶昔封書與君夜，金鑾殿後欲明天。今夜封書在何處？廬山庵裏曉燈前。籠鳥檻猿俱未死，人間相見是何年？微之，微之！此夕此心，君知之乎！"
      ) {
        edges {
          node {
            text
          }
          score
        }
      }
    }`()
    ).toMatchSnapshot();
  });

  afterAll(() => unloadFixtures(fixtures));
});
