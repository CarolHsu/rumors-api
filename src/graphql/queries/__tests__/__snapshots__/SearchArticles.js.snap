exports[`SearchArticles fetches related articles 1`] = `
Object {
  "data": Object {
    "SearchArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "text": "
      憶昔封書與君夜，金鑾殿後欲明天。今夜封書在何處？廬山庵裏曉燈前。籠鳥檻猿俱未死，人間相見是何年？

      微之，微之！此夕此心，君知之乎！
    ",
          },
          "score": 18.429981,
        },
      ],
    },
  },
}
`;
